src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs

```diff
@@ -0,0 +1,392 @@
+            _dictionary = new Dictionary<string, JsonNode>();
+            _duplicatePropertyNameHandling = duplicatePropertyNameHandling;
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonObject"/> class representing provided set of JSON properties.
+        /// </summary>
+        /// <param name="jsonProperties">>Properties to represent as a JSON object.</param>
+        /// <param name="duplicatePropertyNameHandling">Specifies the way of handling duplicate property names.</param>
+        /// <exception cref="ArgumentException">
+        ///   Provided collection contains duplicates if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        /// </exception>
+        public JsonObject(IEnumerable<KeyValuePair<string, JsonNode>> jsonProperties, DuplicatePropertyNameHandling duplicatePropertyNameHandling = default)
+            : this(duplicatePropertyNameHandling)
+            => AddRange(jsonProperties);
+
+        /// <summary>
+        ///   Gets or sets the value of the specified property.
+        /// </summary>
+        /// <param name="propertyName">The property name of the value to get or set.</param>
+        public JsonNode this[string propertyName]
+        {
+            get => GetProperty(propertyName);
+            set => _dictionary[propertyName] = value;
+        }
+
+        /// <summary>
+        ///   Returns an enumerator that iterates through the JSON object properties.
+        /// </summary>
+        /// <returns>An enumerator structure for the JSON object.</returns>
+        /// <exception cref="ArgumentException">
+        ///   Property name to set already exists if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        /// </exception>
+        public IEnumerator<KeyValuePair<string, JsonNode>> GetEnumerator() => _dictionary.GetEnumerator();
+
+        /// <summary>
+        ///   Adds the specified property to the JSON object.
+        /// </summary>
+        /// <param name="jsonProperty">The property to add.</param>
+        /// <exception cref="ArgumentException">
+        ///   Property name to set already exists if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        /// </exception>
+        public void Add(KeyValuePair<string, JsonNode> jsonProperty) => Add(jsonProperty.Key, jsonProperty.Value);
+
+        /// <summary>
+        ///   Adds the specified <see cref="JsonNode"/> property to the JSON object.
+        /// </summary>
+        /// <param name="propertyName">Name of the property to add.</param>
+        /// <param name="propertyValue">Value of the property to add.</param>
+        /// <exception cref="ArgumentException">
+        ///   Property name to set already exists if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        /// </exception>
+        public void Add(string propertyName, JsonNode propertyValue)
```

Comment:
Are we relying on dictionary API to validate for null rather than explicitly validating ourselves? Regardless, add exception docs.

Reply:
I will add a null check and exception docs, thanks.